### PR TITLE
Feature/115 improve dnd

### PR DESCRIPTION
**Pull request by lukefisklennon**

### CODE REVIEW

**Comment by lukefisklennon** [comment]
This PR replaces `react-dnd` with a custom drag-and-drop implementation.

Features include:

- Cross-platform consistency
- Eliminates image flicker
- No longer a need for naive native image clipping
- Better styling (replaces ghosting with depth and shadow effects)
- Animation (when mounting, unmounting and dragging)
- Live state preview as you drag
- Uses intersection area instead of cursor position when dropping (easier to use by Fitts's law)
- Basic touch support (hold-to-drag will be added later)
- Enable mouse wheel scrolling while dragging (auto-scrolling will be added later)

Some issues:

- Due to the performance constraints presented by live preview, and due to styled components [being slow](https://github.com/styled-components/styled-components/issues/2215), a few styled components have been converted to style objects (but are still semantically similar)
- Also, most components are now memoised by custom comparison functions, which may avoid false positives by looking deeper than referential identity; this is something to keep in mind as it may cause bugs with components not re-rendering, so we may want to remove these custom comparators at some point

**Review by maxwowo — CHANGES_REQUESTED** [review]
Damn this is some insane stuff, well freaking done 🔥

File: client/src/components/CourseSelect.tsx [hunk 1]
```diff
@@ -119,6 +117,11 @@ const Weak = styled.span`
   ${weakStyle}
 `;
 
+type MemoProps = {
+  selectedCourses: CourseData[]
+  assignedColors: Record<string, string>
+};
```
[the remark this answers is not in the document]
Comment (lukefisklennon) [reply client/src/components/CourseSelect.tsx:123]: Ah yeah good catch, not sure why that's there

**Review by maxwowo — APPROVED** [review]
Noice
image

**Pull request merged**